mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.kt

```diff
@@ -0,0 +1,1114 @@
+        if (requestWantsWebSockets && responseWantsWebSockets) {
+          handleWebSocketUpgrade(socket, source, sink, request, response)
+          reuseSocket = false
+        } else {
+          writeHttpResponse(socket, sink, response)
+        }
+
+        if (logger.isLoggable(Level.INFO)) {
+          logger.info(
+              "${this@MockWebServer} received request: $request and responded: $response")
+        }
+
+        // See warnings associated with these socket policies in SocketPolicy.
+        when {
+          response.getSocketPolicy() === DISCONNECT_AT_END -> {
+            socket.close()
+            return false
+          }
+          response.getSocketPolicy() === SHUTDOWN_INPUT_AT_END -> socket.shutdownInput()
+          response.getSocketPolicy() === SHUTDOWN_OUTPUT_AT_END -> socket.shutdownOutput()
+          response.getSocketPolicy() === SHUTDOWN_SERVER_AFTER_RESPONSE -> shutdown()
+        }
+        sequenceNumber++
+        return reuseSocket
+      }
+    })
+  }
+
+  @Throws(Exception::class)
+  private fun processHandshakeFailure(raw: Socket) {
+    val context = SSLContext.getInstance("TLS")
+    context.init(null, arrayOf<TrustManager>(UNTRUSTED_TRUST_MANAGER), SecureRandom())
+    val sslSocketFactory = context.socketFactory
+    val socket = sslSocketFactory.createSocket(
+        raw, raw.inetAddress.hostAddress, raw.port, true) as SSLSocket
+    try {
+      socket.startHandshake() // we're testing a handshake failure
+      throw AssertionError()
+    } catch (expected: IOException) {
+      expected.printStackTrace()
+    }
+    socket.close()
+  }
+
+  @Throws(InterruptedException::class)
+  private fun dispatchBookkeepingRequest(sequenceNumber: Int, socket: Socket) {
+    val request = RecordedRequest(
+        "", Headers.of(), emptyList(), 0L, Buffer(), sequenceNumber, socket)
+    requestCount.incrementAndGet()
+    requestQueue.add(request)
+    dispatcher.dispatch(request)
+  }
+
+  /** @param sequenceNumber the index of this request on this connection.*/
+  @Throws(IOException::class)
+  private fun readRequest(
+    socket: Socket,
+    source: BufferedSource,
+    sink: BufferedSink,
+    sequenceNumber: Int
+  ): RecordedRequest? {
+    val request: String
+    try {
+      request = source.readUtf8LineStrict()
+    } catch (streamIsClosed: IOException) {
+      return null // no request because we closed the stream
+    }
+
+    if (request.isEmpty()) {
+      return null // no request because the stream is exhausted
+    }
+
+    val headers = Headers.Builder()
+    var contentLength = -1L
+    var chunked = false
+    var expectContinue = false
+    var header = source.readUtf8LineStrict()
+    while (header.isNotEmpty()) {
+      addHeaderLenient(headers, header)
+      val lowercaseHeader = header.toLowerCase(Locale.US)
+      if (contentLength == -1L && lowercaseHeader.startsWith("content-length:")) {
+        contentLength = header.substring(15).trim { it <= ' ' }.toLong()
```

Comment:
replace `trim { it <= ' ' }` with `trim()`.

Reply:
Fixed.